integration_tests/src/main/python/parquet_test.py

```diff
@@ -779,6 +780,38 @@ def test_spark_32639(std_input_path):
         lambda spark: spark.read.schema(schema_str).parquet(data_path),
         conf=original_parquet_file_reader_conf)
 
+@pytest.mark.skipif(not is_before_spark_320(), reason='Spark 3.1.x does not need special handling')
+def test_spark_40819_31x(std_input_path):
+    data_path = "%s/timestamp-nanos.parquet" % (std_input_path)
+    def read_timestamp_nano_parquet(spark):
+        spark.read.parquet(data_path).collect()
+    # we correctly return timestamp_micros when running against Spark 3.1.x
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: spark.read.parquet(data_path))
+
+@pytest.mark.skipif(is_before_spark_320(), reason='Spark 3.1.x supports reading timestamps in nanos')
```

Comment:
For Spark versions 3.2.0-3.2.4 and 3.3.0-3.3.2, can we add a test where  `spark.sql.legacy.parquet.nanosAsLong=true` and not set at all?